aiohttp/cookiejar.py

```diff
@@ -280,10 +280,25 @@ def filter_cookies(
             if is_not_secure and cookie["secure"]:
                 continue
 
-            # It's critical we use the Morsel so the coded_value
-            # (based on cookie version) is preserved
-            mrsl_val = cast("Morsel[str]", cookie.get(cookie.key, Morsel()))
-            mrsl_val.set(cookie.key, cookie.value, cookie.coded_value)
+            # we need to check, if the found cookie is a better match then the
+            # previously found
+            if name in filtered:
+                existing_cookie = filtered[name]
+                if len(existing_cookie.get("path")) < len(cookie.get("path")):
+                    filtered[name] = cookie
+            else:
+                filtered[name] = cookie
+
+        # It's critical we use the Morsel so the coded_value
+        # (based on cookie version) is preserved
+        for name in filtered:
+            filtered_cookie = filtered[name]
+            mrsl_val = cast(
+                "Morsel[str]", filtered_cookie.get(filtered_cookie.key, Morsel())
+            )
+            mrsl_val.set(
+                filtered_cookie.key, filtered_cookie.value, filtered_cookie.coded_value
+            )
```

Comment:
Can't this whole change be reduced to:
```
if name in filtered and len(cookie["path"]) < len(filtered[name]["path"]):
    continue
```

Alternatively, I'm wondering if we should do it like the spec says and simply order the cookies by path length. At which point we don't need to add the comparisons in here, just take the first cookie that comes through.

Reply:
The spec also suggests that we should actually be sending both cookie values through (e.g. `path-cookie: one; path-cookie zero`).

But, I don't see a way to represent that with SimpleCookie/Morsel.

Reply:
http.cookiejar does appear to do something like this actually:
https://github.com/python/cpython/blob/main/Lib/http/cookiejar.py#L1306

So, e.g. I get a result like (regardless of the order of the 2 cookies) `'Cookie: $Version=1; foo=one; $Path="/bar"; $Port; foo=zero; $Path="/"; $Port'`.

Though, that is using http.cookiejar.Cookie, not http.cookies.SimpleCookie. Looking through, I think the latter is meant more for server-side code, so maybe aiohttp is using the wrong objects here...

Reply:
first of all thanks spending time with this PR

I agree with all of your points - specially the the implementation should follow the spec... But I have to admit, that my current knowledge of python is way to limited in order to provide an spec-valid implementation.

I hope you can agree, that the (based on your first comment adjusted) PR will improve the current situation. For me it was already a though challenge to create this PR (trying to fulfill all the requirements mentioned in https://docs.aiohttp.org/en/stable/contributing.html) - but looks like I made it :-)

If it's better in your opinion to withdraw this PR, create a Issue describing the situation/bug and then work on a complete  solution for the situation this is totally fine for me - using this this PR as some sort of starting point. But then I can't be part in this process (simply cause of my knowledge - and I don't want to break things that had been working for years). So then 'somebody-else' has to provide a valid PR (I don't have to become a contributor).

So it's totally up to the maintainer-team to make a decision here - I came across this (IMHO serious) bug last Friday afternoon when I was working on a migration of an home assistant integration (from 'request' to 'aiohttp') - and it took me the half Saturday to find the root cause - so I wanted to let you know, that 'there is something' (but also already include a 'working-fix').

Reply:
I suspect sending multiple values would need a bigger change, possibly backwards-incompatible, so this is certainly the correct first step.

Reply:
I've noted these improvements in #7583, maybe someone will pick them up.